ext/cl_loader_layers.asciidoc

```diff
@@ -125,6 +137,13 @@ in the <<layer-queries-table, Layer Queries>> table.
 |====
 | Layer Info | Return Type | Description
 | *CL_LAYER_API_VERSION* | cl_layer_api_version | OpenCL Layer API version
+| *CL_LAYER_NAME* | char[] | OpenCL Layer Name
+| *CL_LAYER_OPTIONS* | cl_layer_option[]
+  | OpenCL Layer Options.
+    The list is terminated by a `{NULL, NULL, NULL, NULL}` entry.
+    Layers can, but are not required to, use the requested storage for
+    backing memory. Strings contained in the `cl_layer_option` structure
+    are `NULL` terminated if provided.
```

Comment:
```suggestion
    are `NULL` terminated if provided. Users are required to allocate
    at least the requested amount of memory, and should only free
    this single allocation when they are done using the options.
    Referencing the strings after the user memory has been freed is
    undefined behavior, as the layer may have used the user
    provided allocation for backing storage.
```

Reply:
See here for the expected usage:
https://github.com/KhronosGroup/OpenCL-ICD-Loader/blob/ad9dc9dfc43d25e312d4bfce7536b4316800960f/loader/cllayerinfo.c#L123-L144
```C
    result = p_clGetLayerInfo(CL_LAYER_OPTIONS, 0, NULL, &sz);
    if (CL_SUCCESS == result)
    {
        cl_layer_option *options = (cl_layer_option *)malloc(sz);
        if (options)
        {
            result = p_clGetLayerInfo(CL_LAYER_OPTIONS, sz, options, NULL);
            if (CL_SUCCESS == result)
            {
                size_t i = 0;
                printf("\tCL_LAYER_OPTIONS:\n");
                while (options[i].name)
                {
                    if (options[i].description)
                        printf("\t\t# %s\n", options[i].description);
                    printf("\t\t%s = %s # (default: %s)\n", options[i].name, options[i].value, options[i].default_value);
                    i++;
                }
            }
            free(options);
        }
    }
```
It should be used as any other OpenCL Info query. @bashbaug if you have a better way to formulate it than my clumsy attempt above, I am eager for suggestions.

Reply:
Sorry for the very slow reply.

I frequently use a pattern like this one - the same pattern is also used by the [C++ bindings](https://github.com/KhronosGroup/OpenCL-CLHPP/blob/main/include/CL/opencl.hpp#L1034) for queries that return an array (aka vector) of values:

```c++
        size_t  sizeInBytes = 0;
        clGetPlatformInfo(
            platformId,
            CL_PLATFORM_EXTENSIONS_WITH_VERSION,
            0,
            NULL,
            &sizeInBytes );

        size_t  numExtensions = sizeInBytes / sizeof(cl_name_version);
        std::vector<cl_name_version>  extensions{ numExtensions };
        clGetPlatformInfo(
            platformId,
            CL_PLATFORM_EXTENSIONS_WITH_VERSION,
            sizeInBytes,
            extensions.data(),
            NULL );
```

If the queried "size in bytes" can contain the backing storage for the string also then this pattern won't work.  Are we OK with that?

Reply:
Yes, you are correct and this is why the array is NULL terminated. Considering current options values could be arbitrary, the layers would be hard pressed to guarantee their existence in static memory.